File: Real World Imperative Programming CSharp/DelegateCommand.cs
using System;
using System.Windows.Input;

namespace RealWorldImperativeProgrammingCSharp
{
    public class DelegateCommand : ICommand
    {
        private readonly Action executeAction;
        private readonly Func<bool> canExecute;

        public DelegateCommand(Action executeAction)
            : this(executeAction, () => true)
        {
        }

        public DelegateCommand(Action executeAction, Func<bool> canExecute)
        {
            this.executeAction = executeAction;
            this.canExecute = canExecute;
        }

        public event EventHandler CanExecuteChanged;

        public void Execute(object parameter)
        {
            executeAction();
        }

        public bool CanExecute(object parameter)
        {
            return canExecute();
        }
    }
}

File: Real World Imperative Programming FSharp/DelegateCommand.fs
namespace RealWorldImperativeProgrammingFSharp

open System
open System.Windows.Input


type DelegateCommand (executeAction: Action, canExecute: bool Func) = 

    let canExecuteChanged = new DelegateEvent<EventHandler> ()

    new (executeAction: Action) = DelegateCommand(executeAction, new Func<bool> (fun () -> true))

    interface ICommand with
        [<CLIEvent>]
        member this.CanExecuteChanged = canExecuteChanged.Publish
        member this.Execute _ = executeAction.Invoke ()
        member this.CanExecute _ = canExecute.Invoke ()
